src/mlpack/methods/random_forest/bootstrap.hpp

```diff
@@ -18,30 +18,215 @@ namespace mlpack {
+    for (arma::uword i(0); i < colCount; ++i) {
+      d.Probabilities() = ComputeNextDrawProbabilities(phi, intervals);
+      phi.push_back(d.Random()[0]);
+    }
+
+    return phi;
+  }
+
+  /**
+   * Compute the average uniqueness of each event at any sampling time point.
+   * The average uniqueness is a measure for how isolated an event is during
+   * its lifetime from other events.
+   *
+   * @param[in] intervals Is a `2 x m` matrix, where each column has the start
+   *                      sample and the end sample of an interval.
+   * @param[in] indices Indices of the average uniqueness that is returned.
+   * @return The average uniqueness of the events in @p intervals.
+   */
+  static arma::vec ComputeAverageUniqueness(
+    const IndMatType& intervals,
+    arma::uvec        indices)
+  {
+    arma::vec avg(intervals.n_rows, arma::fill::zeros);
+    arma::uvec concurrency;
+
+    for (arma::uword i(0); i < intervals.n_rows; ++i) {
+      if (i + 1 < intervals.n_rows && intervals(i, 0) > intervals(i + 1, 0)) {
+        throw std::invalid_argument(
+          "intervals must be sorted by starting sample");
+      }
+
+      if (intervals(i, 1) >= concurrency.size()) {
+        concurrency.insert_rows(concurrency.n_rows,
+          1 + intervals(i, 1) - concurrency.n_rows);
+      }
+
+      concurrency.subvec(intervals(i, 0), intervals(i, 1)) += 1;
+    }
+
+    for (arma::uword i(0); i < intervals.n_rows; ++i) {
+      for (arma::uword s(intervals(i, 0)); s <= intervals(i, 1); ++s) {
+        // online averaging algorithm
+        const arma::uword n(s - intervals(i, 0) + 1);
+
+        avg(i) = ((n - 1) * avg(i) + (1.0 / concurrency(s))) / n;
+      }
+    }
+
+    return avg.rows(indices);
+  }
+
+  /**
+   * Compute probabilities of the next draw for each observation.
+   *
+   * @param[in] phi List of previously drawn observations.
+   * @param[in] intervals See ComputeAverageUniqueness().
+   * @return The probabilities for each observation to be drawn in the next
+   *         iteration.
+   */
+  static arma::vec ComputeNextDrawProbabilities(
+    const std::vector<arma::uword>& phi,
+    const IndMatType& intervals)
+  {
+    auto sorter = [&intervals](arma::uword lhs, arma::uword rhs)
+      {
+        return intervals(lhs, 0) < intervals(rhs, 0);
+      };
+    arma::vec avg(intervals.n_rows);
+
+    std::vector<arma::uword> rows(phi);
+
+    std::sort(rows.begin(), rows.end(), sorter);
+
+    rows.reserve(rows.size() + 1);
+    for (arma::uword i(0); i < avg.size(); ++i) {
+      const std::vector<arma::uword>::const_iterator iter(
+        rows.insert(std::upper_bound(rows.begin(), rows.end(), i, sorter),
+          i));
+
+      avg[i] = ComputeAverageUniqueness(
+        intervals.rows(arma::conv_to<arma::uvec>::from(rows)),
+        arma::uvec(1u,
+          arma::fill::value(std::distance(rows.cbegin(), iter)))).back();
+
+      rows.erase(iter);
+    }
+
+    return avg / arma::sum(avg);
```

Comment:
To double check my understanding: here we are returning the average uniqueness of points in each interval.  We will then use this to sample randomly from each interval, is that correct?

Reply:
Here we are returning a probability